### Docking protein-ssRNA 

<div style="text-align: justify;">
This tutorial consists of the following sections:

* **Introduction**
  * Research subject
  * RNA Recognition motifs
  * Single stranded RNA
* Setup/Requirements 
* Installation guide 
* **HADDOCK general concepts**
  * A brief introduction to HADDOCK3
  * Defining restraints for docking
* **Scenario 1. Redocking experimental structure**
  * Extract experimental structure from protein-RNA complex
  * Energy minimisation scoring of experimental models
  * Split the complex structure
  * Defining Restraints from the Protein-RNA binding interface
  * Vizualisation of restraints
  * Docking protein-RNA ensembles from experimental structure
  * Analysis 
* **Scenario 2. Docking Apo-protein and Apo-RNA** 
  * Generate 3D structures
  * SASA scoring for ranking AF3 conformations
  * Extracting the Potential RNA Interface 
  * Docking protocol 
  * Analysis 
* **Scenario 3.Control Test: Catching the Bound Conformation**
  * Edit input files
  * Docking protocol
  * Analysis
* **Conclusions** 
<hr style="border: none; border-top: 2px solid lightgray; width: 100%;">

### Introduction

<p style="text-align: justify;"> This tutorial demonstrates the use of the new modular HADDOCK3 version for predicting the structure of Dead-end protein bound to single-stranded RNA, using its two RNA recognition motifs. Dead-end protein uses RRM1 to bind the RNA, while RRM2, an unusual RRM, works in conjunction with RRM1 to sandwich the RNA between them.

This protein-RNA docking example utilizes web servers to generate structures for the unbound protein and RNA.  Residues for docking will be extracted from the Pesto web server. This approach begins with the protein sequence as input for the AlphaFold web server and the RNA sequence as input for the SimRNA web server. By selecting the top five models generated by each of these tools, we create PDB files containing an ensemble of five structures for both the protein and the RNA as input for docking.</p>

In this tutorial we will be working with  DND1 protein and RNA , PDB code of the complex [7Q4L](https://www.ebi.ac.uk/pdbe/entry/pdb/7q4l) .

<hr style="border: none; border-top: 1.2px solid lightgray; width: 100%;">

### Research subject
<p style="text-align: justify;">The structure of Dead End protein (PDB ID 7Q4L) bound to RNA was determined using NMR spectroscopy. This protein plays a crucial role in regulating gene expression by controlling mRNA breakdown. It binds to a specific RNA sequence, forming a unique interaction. This binding involves a precise fit between the protein and RNA, stabilized by hydrogen bonds and other interactions, such as van der Waals forces and hydrophobic interactions. The protein's conformation also changes slightly to accommodate the RNA, further strengthening the binding. These intricate interactions ensure that the protein binds to the correct RNA sequence and effectively regulates gene expression.
</p>


 <img src="/figures/course/7q4l_complex.png" alt="A descriptive alt text for your image" width="75%">

<hr style="border: none; border-top: 1.2px solid lightgray; width: 100%;">

### RNA recognition motifs 
<p style="text-align: justify;"> RNA recognition motifs (RRMs) are versatile protein domains that play a pivotal role in recognizing and binding RNA molecules. They are characterized by a conserved structure consisting of four beta-strands and two alpha-helices. The RRM's ability to recognize RNA stems from its capacity to form specific interactions with the RNA backbone and individual nucleotides.</p>


<img src="/figures/course/protein_rrms.png" alt="A descriptive alt text for your image" width="75%">

<hr style="border: none; border-top: 1.2px solid lightgray; width: 100%;">

### Single stranded RNA 
<p style="text-align: justify;"> The RNA bound to the Dead End protein (PDB ID 7Q4L) is a relatively short, single-stranded molecule. While it does not form complex secondary structures like hairpins or loops, it adopts a specific conformation upon binding to the protein. This induced-fit model, revealed by the 7Q4L structure, shows that the RNA molecule adapts its shape to optimize interactions with the protein. The RNA's flexibility allows it to bend and twist, forming specific contacts with the protein's amino acid side chains. </p>


<img src="/figures/course/rna_7q4l.png" alt="A descriptive alt text for your image" width="75%">
<br>
<hr style="border: none; border-top: 2px solid lightgray; width: 100%;">

### Setup/Requirements

In order to have all the files from this tutorial , including already run docking protocols download the .zip file: [download TUTORIAL](https://github.com/roxanavas/protein_rna_tutorial/archive/refs/heads/main.zip)

* `haddock3`: Contains HADDOCK3 configuration and job files for the various scenarios in this tutorial
* `pdbs`: Contains the pre-processed PDB files
* `restraints`: Contains the interface information and the correspond restraint files for HADDOCK
* `docking_results`: Contains pre-calculated run results for the various scenarios in this tutorial
* `emscoring`: Contains all the emscoring run results 
* `sasa_scoring`: Contains all the SASA run results
* `scripts`: Contains a variety of scripts used in this tutorial
* `vmd_scripts`: Special scripts for visualisation with VMD
* `AF3_results`: Contains all the AF3 output files 
* `simrna_results`: Contains all the SimRNA output files

<hr style="border: none; border-top: 2px solid lightgray; width: 100%;">

### Installation guide 

First thing to do after creating an account for NMRBOX --> [CREATE NEW ACCOUNT](https://nmrhub.org/register?returnUrl=https:%2F%2Fnmrbox.nmrhub.org%2Fsignup), check out their [GETTING STARTED](https://nmrbox.nmrhub.org/pages/getting-started) tutorial to have you virtual machine setup ready. 
On this virtual machine we will need a few things to be installed. 
* `Miniconda`: Download and install Miniconda 
    
       wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh -O miniconda.sh && bash miniconda.sh
    Activate miniconda3:     `source miniconda3/bin/activate`


* `Haddock3`: Install Haddock3 with miniconda 

  * Create and activate new haddock environment:

        conda create -n haddock3 python=3.9
        conda activate haddock3

  * Install HADDOCK3:

        pip install haddock3 

  *  **CNS** setup 

          wget -O cns.linux https://surfdrive.surf.nl/files/index.php/s/sJg6mmzHWES4x1o/download
     * copy binary file in haddock 

           cp <my-cns-binary> /home/testuser/.venv/lib/python3.9/site-packages/haddock/bin/cns

* `pymol`: Install pymol package for a script we use later on 


        conda install -c conda-forge pymol-open-source

* `Haddock-restraints`:

       wget -qO- "https://github.com/haddocking/haddock-restraints/releases/download/v0.8.1/haddock-restraints-v0.8.1-x86_64-unknown-linux-musl.tar.gz" | tar -xz


* `VMD`: it is already installed on NMRBOX machines

<hr style="border: none; border-top: 2px solid lightgray; width: 100%;">

### HADDOCK general concepts

HADDOCK (see [https://www.bonvinlab.org/software/haddock2.4](https://www.bonvinlab.org/software/haddock2.4))
is a collection of python scripts derived from ARIA ([https://aria.pasteur.fr](https://aria.pasteur.fr))
that harness the power of CNS (Crystallography and NMR System – [https://cns-online.org](https://cns-online.org))
for structure calculation of molecular complexes. What distinguishes HADDOCK from other docking software is its ability,
inherited from CNS, to incorporate experimental data as restraints and use these to guide the docking process alongside
traditional energetics and shape complementarity. Moreover, the intimate coupling with CNS endows HADDOCK with the
ability to actually produce models of sufficient quality to be archived in the Protein Data Bank.

A central aspect to HADDOCK is the definition of Ambiguous Interaction Restraints or AIRs. These allow the
translation of raw data such as NMR chemical shift perturbation or mutagenesis experiments into distance
restraints that are incorporated in the energy function used in the calculations. AIRs are defined through
a list of residues that fall under two categories: active and passive. Generally, active residues are those
of central importance for the interaction, such as residues whose knockouts abolish the interaction or those
where the chemical shift perturbation is higher. Throughout the simulation, these active residues are
restrained to be part of the interface, if possible, otherwise incurring in a scoring penalty. Passive residues
are those that contribute for the interaction, but are deemed of less importance. If such a residue does
not belong in the interface there is no scoring penalty. Hence, a careful selection of which residues are
active and which are passive is critical for the success of the docking.

<hr style="border: none; border-top: 1.2px solid lightgray; width: 100%;">

### A brief introduction to HADDOCK3


HADDOCK3 is the next generation integrative modelling software in the
long-lasting HADDOCK project. It represents a complete rethinking and rewriting
of the HADDOCK2.X series, implementing a new way to interact with HADDOCK and
offering new features to users who can now define custom workflows.

In the previous HADDOCK2.x versions, users had access to a highly
parameterisable yet rigid simulation pipeline composed of three steps:
`rigid-body docking (it0)`, `semi-flexible refinement (it1)`, and `final refinement (itw)`.

<figure style="text-align: center;">
<img width="75%" src="https://www.bonvinlab.org/education/HADDOCK3/HADDOCK3-antibody-antigen/HADDOCK2-stages.png">
</figure>

In HADDOCK3, users have the freedom to configure docking workflows into
functional pipelines by combining the different HADDOCK3 modules, thus
adapting the workflows to their projects. HADDOCK3 has therefore developed to
truthfully work like a puzzle of many pieces (simulation modules) that users can
combine freely. To this end, the “old” HADDOCK machinery has been modularized,
and several new modules added, including third-party software additions. As a
result, the modularization achieved in HADDOCK3 allows users to duplicate steps
within one workflow (e.g., to repeat twice the `it1` stage of the HADDOCK2.x
rigid workflow).

Note that, for simplification purposes, at this time, not all functionalities of
HADDOCK2.x have been ported to HADDOCK3, which does not (yet) support NMR RDC,
PCS and diffusion anisotropy restraints, cryo-EM restraints and coarse-graining.
Any type of information that can be converted into ambiguous interaction
restraints can, however, be used in HADDOCK3, which also supports the
*ab initio* docking modes of HADDOCK.

<figure style="text-align: center;">
<img width="75%" src="https://www.bonvinlab.org/education/HADDOCK3/HADDOCK3-antibody-antigen/HADDOCK3-workflow-scheme.png">
</figure>

To keep HADDOCK3 modules organized, we catalogued them into several
categories. But, there are no constraints on piping modules of different
categories.

The main module categories are "topology", "sampling", "refinement",
"scoring", and "analysis". There is no limit to how many modules can belong to a
category. Modules are added as developed, and new categories will be created
if/when needed. You can access the HADDOCK3 documentation page for the list of
all categories and modules. Below is a summary of the available modules:

* **Topology modules**
    * `topoaa`: *generates the all-atom topologies for the CNS engine.*
* **Sampling modules**
    * `rigidbody`: *Rigid body energy minimization with CNS (`it0` in haddock2.x).*
    * `lightdock`: *Third-party glow-worm swam optimization docking software.*
* **Model refinement modules**
    * `flexref`: *Semi-flexible refinement using a simulated annealing protocol through molecular dynamics simulations in torsion angle space (`it1` in haddock2.x).*
    * `emref`: *Refinement by energy minimisation (`itw` EM only in haddock2.4).*
    * `mdref`: *Refinement by a short molecular dynamics simulation in explicit solvent (`itw` in haddock2.X).*
* **Scoring modules**
    * `emscoring`: *scoring of a complex performing a short EM (builds the topology and all missing atoms).*
    * `mdscoring`: *scoring of a complex performing a short MD in explicit solvent + EM (builds the topology and all missing atoms).*
* **Analysis modules**
    * `caprieval`: *Calculates CAPRI metrics (i-RMSD, l-RMSD, Fnat, DockQ) with respect to the top scoring model or reference structure if provided.*
    * `clustfcc`: *Clusters models based on the fraction of common contacts (FCC)*
    * `rmsdmatrix`: *Calculates the pairwise RMSD matrix between all the models generated in the previous step.*
    * `ilrmsdmatrix`: *Calculates the pairwise interface-ligand RMSD matrix between all the models generated in the previous step.*
    * `clustrmsd`: *Clusters models based on pairwise RMSD matrix calculated with the `rmsdmatrix`/`ilrmsdmatrix` module.*
    * `seletop`: *Selects the top N models from the previous step.*
    * `seletopclusts`: *Selects top N clusters from the previous step.*
    * `contactmap`: *Generates a contact map for the models generated in the previous step.*
    * `alascan`: *Performs an alanine scanning on the models generated in the previous step.*

The HADDOCK3 workflows are defined in simple configuration text files, similar to the TOML format but with extra features.
Contrarily to HADDOCK2.X which follows a rigid (yet highly parameterisable)
procedure, in HADDOCK3, you can create your own simulation workflows by
combining a multitude of independent modules that perform specialized tasks.

<hr style="border: none; border-top: 2px solid lightgray; width: 100%;">

### Defining restraints for docking

Before setting up the docking we need first to generate distance restraint files
in a format suitable for HADDOCK.  HADDOCK uses [CNS](https://cns-online.org/) as computational
engine. A description of the format for the various restraint types supported by
HADDOCK can be found in our [Nature Protocol](https://www.nature.com/articles/s41596-024-01011-0.epdf?sharing_token=UHDrW9bNh3BqijxD2u9Xd9RgN0jAjWel9jnR3ZoTv0O8Cyf_B_3QikVaNIBRHxp9xyFsQ7dSV3t-kBtpCaFZWPfnuUnAtvRG_vkef9o4oWuhrOLGbBXJVlaaA9ALOULn6NjxbiqC2VkmpD2ZR_r-o0sgRZoHVz10JqIYOeus_nM%3D) paper, Box 4.

Distance restraints are defined as:

<pre style="background-color:#DAE4E7">
assign (selection1) (selection2) distance, lower-bound correction, upper-bound correction
</pre>

The lower limit for the distance is calculated as: distance minus lower-bound
correction and the upper limit as: distance plus upper-bound correction.  The
syntax for the selections can combine information about chainID - `segid`
keyword -, residue number - `resid` keyword -, atom name - `name` keyword.
Other keywords can be used in various combinations of OR and AND statements.
Please refer for that to the [online CNS manual](http://cns-online.org/v1.3/).

We will shortly explain in this section how to generate both ambiguous
interaction restraints (AIRs) and specific distance restraints for use in
HADDOCK illustrating two scenarios:

* **Haddock-restraints interface extracted from the 7q4l complex within a distance of 4.5 A**
* **Pesto webserver using the dead-end protein**

<hr style="border: none; border-top: 2px solid lightgray; width: 100%;">

### Scenario 1. Redocking experimental structure

In this experiment, we use HADDOCK3 to investigate the binding mode of the DND1 protein in complex with a single stranded RNA, as determined experimentally and extracted from the PDB database. The PDB file contains the atomic coordinates of the complex, providing a detailed structural model with precise spatial positions of all atoms.
We will extract the protein and RNA components from the PDB structure, preserving their experimentally determined bound conformation. Using these two structures, we will perform a docking workflow (outlined below) to simulate their interaction and assess the docking results.



<img src="/figures/course/Scenario_1_workflow.png" alt="A descriptive alt text for your image" width="95%">
<br>
<hr style="border: none; border-top: 1.2px solid lightgray; width: 100%;">

### 1. Extract experimental structure from protein-RNA complex

    pdb_fetch 7q4l | pdb_keepcoord |pdb_delhetatm |pdb_reres -1 > reference.pdb

Let's examine the reference; the PDB file contains 20 models, from which we need to select the best 5 candidates for our ensemble redocking protocol. This selection can be made by running an EMscoring in HADDOCK3.

<hr style="border: none; border-top: 1.2px solid lightgray; width: 100%;">

### 2. Energy minimisation scoring of experimental models 

EMscoring in HADDOCK3 evaluates the quality of conformations by scoring them based on electrostatic and van der Waals interactions, as well as solvent accessibility. This helps rank the conformations and select the most energetically favorable poses for further analysis.
   
    # ====================================================================
    # directory in which the scoring will be done
    run_dir = "emscoring_reference"
    
    # execution mode
    ncores = 30
    mode = "local"
    
    # Self contained rundir (to avoid problems with long filename paths)
    self_contained = true

    # ensemble of different complexes to be scored
    molecules = ["reference.pdb"]
    
    # ====================================================================
    # Parameters for each stage are defined below
    
    [topoaa]
    
    [emscoring]
    tolerance = 20
    per_interface_scoring = true

    # ====================================================================
The table with the emscoring can be found in the docking directory ~/emscoring_reference/1_emscoring/emscoring.tsv


<img src="/figures/course/emscore_table.png" alt="A descriptive alt text for your image" width="75%">
<br>

<hr style="border: none; border-top: 1.2px solid lightgray; width: 100%;">

### 3. Split the complex structure
* Selection of top 5 ranked conformations based on EM Scoring

      pdb_selmodel -19,16,11,9,20 reference.pdb > reference_5.pdb

* Selection of top 5 protein and RNA models from complex

      pdb_selchain -A reference.pdb |pdb_selmodel -19,16,11,9,20 > protein_5.pdb

      pdb_selchain -B reference.pdb |pdb_selmodel -19,16,11,9,20 > rna_5.pdb

<hr style="border: none; border-top: 1.2px solid lightgray; width: 100%;">

### 4. Defining Restraints from the Protein-RNA binding interface

* extract the interface using **haddock-restraints**, the reference file is needed and the cutoff for the interface.

    haddock-restraints interface reference.pdb 4.5

The output shows the residue numbers for chain A (protein) and chain B(RNA) that are present in the interface. 

    Chain A: [1, 2, 28, 29, 30, 31, 51, 52, 53, 55, 56, 80, 82, 83, 84, 85, 86, 87, 88, 90, 91, 92, 93, 94, 95, 96, 107, 111, 112, 120, 122, 123, 124, 125, 126, 127, 128, 129, 181, 182, 185, 186, 188, 189, 192, 193, 206, 207, 209, 210, 211]
    Chain B: [228, 229, 230, 231, 232, 233, 234, 235]

The extracted residues from the specified interface will be added in the extracted_interface_4.5.json

    cp ~/restraints/default.json   extracted_interface_4.5.json

* insert the residues from **haddock-restraints** for chain A as active and passive for chain B.

        "id": 1,
        "chain": "A",
        "active": [1, 2, 28, 29, 30, 31, 51, 52, 53, 55, 56, 80, 82, 83, 84, 85, 86, 87, 88, 90, 91, 92, 93, 94, 95, 96, 107, 111, 112, 120, 122, 123, 124, 125, 126, 127, 128, 129, 181, 182, 185, 186, 188, 189, 192, 193, 206, 207, 209, 210, 211],
    
        "id": 2,
        "chain": "B",
        "passive": [228, 229, 230, 231, 232, 233, 234, 235],

## 
Generate the restraints file using the residues as active residues for the docking.

    haddock-restraints tbl extracted_interface_4.5.json > active_residues_4.5.tbl

<hr style="border: none; border-top: 1.2px solid lightgray; width: 100%;">

### Visualization of restraints with VMD

Residue visualization is essential to confirm that only those directly involved in binding are considered. Notably, the residues vary across the five structures analyzed, reflecting structural dynamics. In particular, the shifting tail of RRM1 appears to influence the residues involved in the binding.
<br> The VMD script used is in ~/vmd_scripts/view_restraints.tcl .

<br>


<img src="/figures/course/haddock_restraints.gif" alt="A descriptive alt text for your image" width="75%">
<br>

<hr style="border: none; border-top: 1.2px solid lightgray; width: 100%;">

### 5. Docking protein-RNA ensembles from experimental structure 

For the docking we will need our protein_5.pdb rna_5.pdb and reference_5.pdb and the restraints file generated before and the docking configuration files with all the steps and parameters.

      # ====================================================================
      #     Redocking experimental structure 7Q4L, top 5 models
      # ====================================================================
      
      # directory name of the run
      run_dir = "redocking_ens_5"
      
      # local mode
      mode = "local"
      ncores = 50
      # Self contained rundir (to avoid problems with long filename paths)
      self_contained = true
      
      # molecules to be docked
      molecules =  [
                    "protein_5.pdb",
                    "rna_5.pdb"
      ]
      
      # ====================================================================
      # Parameters for the various stages
      # ====================================================================
      [topoaa]
      
      [rigidbody]
      # surface ambig restraints
      ambig_fname = "active_residues_4.5.tbl"
      # Number of models to generate
      sampling = 10000
      
      [clustfcc]
      min_population = 4
      
      [seletopclusts]
      ## select all the clusters
      top_cluster = 5
      ## select the best 10 models of each cluster
      top_models = 10
      
      [caprieval]
      reference_fname = "reference_5.pdb"
      allatoms = True
      
      [flexref]
      # Acceptable percentage of model failures
      tolerance = 5
      ambig_fname = "active_residues_4.5.tbl"
      dnarest_on=True
      
      [emref]
      # surface ambig restraints
      ambig_fname = "active_residues_4.5.tbl"
      
      [clustfcc]
      
      [seletopclusts]
      top_cluster = 5
      
      [caprieval]
      reference_fname = "reference_5.pdb"
      allatoms = True
      
      # ====================================================================
           
Run your first **HADDOCK3** protocol 

    haddock3 redocking_5_ens.cfg

<hr style="border: none; border-top: 1.2px solid lightgray; width: 100%;">

### 6. Analysis of results based on CAPRI analysis
CAPRI (Critical Assessment of PRedicted Interactions) is a crucial, community-driven evaluation platform for protein-protein docking methods like HADDOCK3.  It provides a standardized way to assess the accuracy of HADDOCK3's predictions by comparing them to experimentally determined structures.

The workflow output includes:
*    **Log** file: Run information, including timings at the end.
*    **analysis** directory:** Plots and a comprehensive report (report.html).
*    **data** directory: Input PDB and restraint files.
*    **toppar** directory: Force field files (self-contained mode only).
*    **traceback** directory: traceback.tsv links models across workflow steps.
*    Module directories (e.g., 08_seletopclusts): Output PDB files. Subdirectories (e.g., cluster_1) contain ranked clusters, with origin information in a text file (e.g., seletopclusts.txt).
*    **XX_caprieval** directories: capri_ss.tsv file with model names, rankings, and statistics (score, iRMSD, Fnat, lRMSD, ilRMSD, DockQ). These are the best source for ranking and HADDOCK score information.


<img src="/figures/course/capri_table.png" alt="A descriptive alt text for your image" width="75%">
<br>

Insights on the other score in the CAPRI analysis like HADDOCK3 score, i-rmsd, DOCkQ.

`capri_ss.tsv` file contains model-specific statistics:
* `score`: HADDOCK score.
* `irmsd`: Interface RMSD.
*  `fnat`: Fraction of native contacts.
*  `lrmsd`: Ligand RMSD.
*  `ilrmsd`: Interface-ligand RMSD.
*  `dockq`: DockQ score (0-1 scale).
*  Other terms: `bsa` (buried surface area), `elec` (electrostatic energy), `vdw` (van der Waals energy), `desolv` (desolvation energy).

CAPRI quality definitions:

*    **Acceptable:** i-RMSD < 4Å or l-RMSD < 10Å and Fnat > 0.1 (0.23 < DOCKQ < 0.49)
*    **Medium:** i-RMSD < 2Å or l-RMSD < 5Å and Fnat > 0.3 (0.49 < DOCKQ < 0.8)
*    **High:** i-RMSD < 1Å or l-RMSD < 1Å and Fnat > 0.5 (DOCKQ > 0.8)

Cluster statistics  are in capri_clt.tsv:

*    **cluster_rank**: Cluster rank.
*    **cluster_id**: Cluster ID (1 = largest).
*    **n**: Number of models in the cluster.
*    Averages and standard deviations for score, iRMSD, Fnat, lRMSD, DockQ, etc.

HTML reports in the analysis/XX_caprieval directories provide a user-friendly way to view these results.


<img src="/figures/course/capri_plots.png" alt="A descriptive alt text for your image" width="95%">
<img src="/figures/course/capri_plots_2.png" alt="A descriptive alt text for your image" width="95%">
<br>

After analyzing the capri table and plots, it's crucial to examine the structure. The image shows the top-ranked docking model superimposed on the 1st scored reference model.

<img src="/figures/course/redocked_vs_ref.png" alt="A descriptive alt text for your image" width="75%">
<br>
<hr style="border: none; border-top: 2px solid lightgray; width: 100%;">


### Scenario 2. Docking Apo-protein and Apo-RNA 

After achieving good results with redocking, we now face the challenge of building a protein-RNA complex without prior structural data. Using only the **FASTA** files for the RNA and protein, we can predict their 3D structures.
In this scenario, we will show you the capabilities and challenges of using computational tools, including web-based servers, to generate these structures and perform docking to form a protein-RNA complex.

<br>


<img src="/figures/course/Scenario_2_workflow.png" alt="A descriptive alt text for your image" width="95%">
<br>
<hr style="border: none; border-top: 1.2px solid lightgray; width: 100%;">

### 1. Generate 3D structures
**1.1. Generate RNA conformations**  <br>
SimRNA is a computational tool that simulates RNA folding and predicts its 3D structure. It simplifies the complex RNA molecule into a coarse-grained representation, reducing computational cost. By using statistical potentials and Monte Carlo simulations, SimRNA explores the possible conformations of RNA and selects the most likely structure. Additional experimental data, like secondary structure predictions, can be incorporated to refine the model.
This step is done on the **SimRNAweb.2.0**  with the default settings on the webserver, the input for the run is just the RNA sequence from the 7Q4L experiment in the PDB database from the FASTA file. 

    >7Q4L_2|Chain B|RNA (5'-R(*CP*UP*UP*AP*UP*UP*UP*G)-3')|Homo sapiens (9606)
    CUUAUUUG


As our RNA is single stranded and we don't want any base-pairing in our structure in SimRNA webserver we will provide:

    Hard secondary structure restraints:
      xxxxxxxx

The SimRNA  webserver: https://genesilico.pl/SimRNAweb/ <br>
For the tutorial, if there are any issues when running the webserver the results generated by us are provided in ~/simrna_results/ .


Create a directory named simrna_results/ and copy all the generated PDB files containing the RNA structures into it. Then, execute the following command. This command will generate an ensemble, update the chain identifier from "A" to "B" (since chain A is reserved for the protein in our complex), and renumber the residues starting from 228. This renumbering helps avoid overlap with the protein residues when generating the HADDOCK restraints file.


    pdb_mkensemble rna_7q4l-3fca3605_sim_thrs0.80A_*  |pdb_rplchain  -A:B |pdb_reres  -228 > simrna_5.pdb

<hr style="border: none; border-top: 1.2px solid lightgray; width: 100%;">

**1.2. Generate protein conformations**  <br>
**AF3** is a webserver that uses the advanced AlphaFold3 model to predict protein structures with high accuracy. Simply submit an amino acid sequence through its easy-to-use interface, and the server processes it on remote hardware to deliver a detailed 3D model. This simplicity and reliability make AF3 a valuable tool for protein research and drug discovery.

The input protein sequence for AF3:

    >7Q4L_1|Chain A|Dead end protein homolog 1|Homo sapiens (9606)
    GAMERVNPENKAALEAWVRETGIRLVQVNGQRKYGGPPPGWVGSPPPAGSEVFIGRLPQDVYEHQLIPLFQRVGRLYEFRLMMTFSGLNRGFAYARYSSRRGAQAAIATLHNHPLRPSCPLLVCRSTEKCELSVDGLPPNLTRSALLLALQPLGPGLQEARLLPSPGPAPGQIALLKFSSHRAAAMAKKALVEGQSHLCGEQVAVEWLKPDLKQRLRQQLVGPFLRS

AF3 webserver: https://alphafoldserver.com/

<img src="/figures/course/af3_web.png" alt="A descriptive alt text for your image" width="100%">
<br>

**pLDDT** (Predicted Local Distance Difference Test) is a confidence score ranging from 0 to 100, indicating how reliable the local structure prediction is for each residue. According to the AlphaFold Protein Structure Database:

*    90-100 → High accuracy, well-structured region
*    70-90 → Good backbone prediction, generally reliable
*    50-70 → Low confidence, should be interpreted with caution
*    <50 → Very low confidence, likely disordered and unreliable

**PAE** (Predicted Aligned Error) measures the **expected positional error** between residue pairs and is visualized as a 2D plot.It indicates how confidently AlphaFold predicts the relative positioning of different residues or domains.

* Low PAE values suggest that AlphaFold confidently predicts the relative positioning of residues or domains.
* High PAE values indicate uncertainty in domain orientation, meaning their positions relative to each other should not be interpreted as definitive.

For this case, we performed five separate AlphaFold experiments using different random seeds ranging from 10 to 1000, and labeled them as follows:
  * protein_e1_t1_10
  * protein_e1_t2_50
  * protein_e1_t3_100
  * protein_e1_t4_500
  * protein_e1_t5_1000 </br>
Select all 5 runs and save them as a zip file from the AF3 webserver.
These files need some processing, unzip all files in a new directory AF3_results/


         mkdir -p AF3_results && unzip ~/Downloads/fold_protein.zip  -d AF3_results


In this  directory we will find the 5 directories with the AF3 results in .cif format that need to be tranformed in an ensemble pdb file.
In order to do this, use the cif_to_pdb.py from ~/scripts/cif_to_pdb.py. This script is looking inside all AF3_results directories and generates the .pdb files from all .cif files.

    python ~/scripts/cif_to_pdb.py ~/path_to/AF3_results
After all AF3 results are in pdb format we can generate the ensemble using the command :

      pdb_mkensemble protein_e1_t1_10/*.pdb protein_e1_t2_50/*.pdb protein_e1_t3_100/*.pdb protein_e1_t4_500/*.pdb protein_e1_t5_1000/*.pdb  > af_ens.pdb

<hr style="border: none; border-top: 1.2px solid lightgray; width: 100%;">

### 2. SASA scoring for ranking AF3 conformations
Previous research on RRMS proteins has shown that the β-sheet region plays a key role in RNA binding, as the amino acids in this region are primarily responsible for interacting with RNA. Since we need five structures for ensemble docking with RNA, we'll pick the ones with the best shape complementarity to make complex formation easier. To do this, we'll first extract the β-sheet residues from the structure using VMD.

    vmd -e extract_betasheet.tcl 

Next, we'll calculate **surface accessibility scores** for all AlphaFold conformations using the HADDOCK3 protocol. Based on these scores, the top five conformations will be selected for further docking.
The SASA score is calculated by evaluating the accessibility of residues. The score represents the sum of residues that are categorized as either buried or accessible, and how well each residue fits into the correct category.

The accessibility score is calculated as:

* **buried residues** are those that are not exposed to solvent, typically involved in hydrophobic interactions.
* **accessible residues** are those that are exposed to the solvent, often involved in hydrogen bonding or other interactions with the environment.

* manually select the residues from **betasheet_resids.txt** insert into ~/haddock3/sasa_scoring_af.cfg 
     
          # ====================================================================
          # directory in which the scoring will be done
          run_dir = "sasa_scoring_af"
    
          # execution mode
          ncores = 50
          mode = "local"
          # Self contained rundir (to avoid problems with long filename paths)
          self_contained = true
    
          # ensemble of different complexes to be scored
          molecules = ["af_ens.pdb",]
          # ====================================================================
          # Parameters for each stage are defined below
        
          [topoaa]
        
          [sasascore]
          resdic_accessible_A =["insert you residues here"]
          probe_radius = 2.5
        
          # ====================================================================

Run the SASA scoring protocol using the following command:
      
      haddock3 sasa_scoring_af.cfg


Once the protocol has finished, you can visualize the results as shown below:
<br>

<img src="/figures/course/sasa_score.png" alt="A descriptive alt text for your image" width="95%">
<br>
This image illustrates how the SASA score increases from 53, with the domains far apart, to 60 when the RNA recognition domains come closer, reducing the available space for RNA binding.
<br>
<br>

<img src="/figures/course/str_sasa_scoring.png" alt="A descriptive alt text for your image" width="95%">
<br>
<br>
 
After running the script we extract the top 5 structures and generate an ensemble file for the docking.

    pdb_mkensemble fold_t1_7q4l_10_model_2.pdb fold_t1_7q4l_10_model_3.pdb fold_t2_7q4l_20_model_2.pdb fold_t1_7q4l_10_model_1.pdb fold_t1_7q4l_10_model_0.pdb > af3_5.pdb

<hr style="border: none; border-top: 1.2px solid lightgray; width: 100%;">

### 3. Extracting the Potential RNA Interface 

Residue selection for HADDOCK is guided by **PeSTo**, which predicts RNA-binding sites using the ensemble structure af_5.pdb as input. The webserver visualizes residues colored by their predicted probability of RNA interaction. We prioritize those with high scores (>0.80) and ensure spatial distribution across the protein surface to capture the potential binding interface.

<br>
<img src="/figures/course/pesto_structure.png" width="100%">
<br>

From the **PeSTo** table we chose the residues with a value > 0.80, we extract them manually and input in a json file for the haddock-restraints to generate the ambiguos file for the docking. 
* A small issue here is that **PeSTo** gives different resid number as in our structure the protein starts from resid 9 not 1 we have to make sure the aminoacids have the same ids in our pdb for the docking. 

We download the **PeSTo** file from the webserver and extract the residues based on the occupancy in the **pesto** output. 
To extract the list of residues we run the script. The output from pesto is a pdb file that in webserver issues can be found in ~/pdbs/

    python ~/scripts/extract_residues_pesto.py pesto_prediction.pdb 0.80    

Output : 

    These are all residues : [3, 4, 27, 28, 29, 30, 31, 49, 51, 53, 55, 56, 57, 59, 80, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 94, 111, 112, 118, 120, 121, 122, 123, 124, 125, 126, 127, 129, 185, 186, 189]

* create a new pesto_res.json file with the new residues.

      cp default.json pesto_res.json

* insert the residues from **PeSTo** into the active chain A , and add the RNA residue numbers as passive chain B 


        "id": 1,
        "chain": "A",
        "active": [1, 2, 3, 4, 27, 28, 30, 53, 55, 56, 59, 83, 84, 85, 86, 88, 89, 90, 91, 92, 94, 118, 119, 120, 182, 185],
    
        "id": 2,
        "chain": "B",
        "passive": [228, 229, 230, 231, 232, 233, 234, 235],


The residues will be processed in the **haddock-restraints** to get the restraints file for the docking. 

    haddock-restraints tbl pesto_restraints.json  > pesto_residues_80.tbl

<hr style="border: none; border-top: 1.2px solid lightgray; width: 100%;">

### 4. Docking protocol 
To dock the apo protein and apo RNA conformations, the protocol will be adjusted. One key change from our initial redocking protocol is setting **randremoval** to False. Previously, this option randomly removed residues to try to improve the structure’s energy pose. However, since the generated structures don’t have bound conformations, we need to keep all active residues in the docking process. <br>
In this scenario, CAPRI evaluation is performed without a reference structure, so the best docking complex is ranked solely based on its lowest energy score.

      # ====================================================================
      #    Docking of ens 5 AF protein and ens 5 simrna
      # ====================================================================
      
      # directory name of the run
      run_dir = "apo_ens_5"
      
      # local mode
      mode = "local"
      ncores = 50
      # Self contained rundir (to avoid problems with long filename paths)
      self_contained = true
      
      # molecules to be docked
      molecules =  [
      "af_5.pdb",
      "simrna_5.pdb"
      ]
      
      # ====================================================================
      # Parameters for the various stages
      # ====================================================================
      [topoaa]
      
      [rigidbody]
      # surface ambig restraints
      ambig_fname = "pesto_residues_80.tbl"
      # Number of models to generate
      sampling = 10000
      randremoval = False
      
      [clustfcc]
      min_population = 4
      
      [seletopclusts]
      ## select all the clusters
      top_cluster = 10
      ## select the best 10 models of each cluster
      top_models = 10
      
      [caprieval]
      allatoms = True
      
      [flexref]
      # Acceptable percentage of model failures
      tolerance = 5
      ambig_fname = "pesto_residues_80.tbl"
      dnarest_on=True
      randremoval = False
      
      [emref]
      # surface ambig restraints
      ambig_fname = "pesto_residues_80.tbl"
      randremoval = False
      
      [clustfcc]
      
      [seletopclusts]
      top_cluster = 10
      
      [caprieval]
      allatoms = True
      
      [caprieval]
      reference_fname = "reference_5.pdb"
      allatoms = True
      
      # ====================================================================

<hr style="border: none; border-top: 1.2px solid lightgray; width: 100%;">

### 5. Analysis of docked complexes
Additionally, we perform an extra CAPRI evaluation by adding the reference structure, allowing us to compare the results with the initial experiment. These evaluations will help highlight the challenges of determining a complex structure from apo protein and apo RNA using solely computational methods.


<img src="/figures/course/apo_capri_table.png" width="75%">
<img src="/figures/course/apo_capri_plots.png" width="95%">
<img src="/figures/course/apo_capri_plots_2.png" width="95%">
<br>

Visualisation of the new results , 1st ranked model by caprieval in HADDOCK3 superimposed to the best ranked reference from experimental data.

<br>

<img src="/figures/course/apo_vs_ref.png" width="95%">
<br>

<hr style="border: none; border-top: 2px solid lightgray; width: 100%;">

### Scenario 3. Control Test: Catching the Bound Conformation
For this we will need some editing to our files , so that we include in the input structures 1 protein and 1 RNA in bound conformations. This way we will get a new ensemble of top 4 AF and top 4 Simrna adding the complex structures as the 5th conformation to be docked. In order to do this let's first edit the pdb input files.
### 1. Edit input files
* new protein ensemble 
  
      pdb_selmodel -2,3,4,18 af_5.pdb > af_4.pdb
      pdb_selmodel -9 protein_5.pdb > protein_1.pdb
      pdb_merge protein_1.pdb af_4.pdb |pdb_keepcoord > af_4_p_1.pdb

* new RNA ensemble 
   
      pdb_selmodel -1,2,3,4 simrna_5.pdb > simrna_4.pdb
      pdb_selmodel -9 rna_5.pdb > rna_1.pdb
      pdb_merge rna_1.pdb simrna_4.pdb |pdb_keepcoord > simrna_4_r_1.pdb

<hr style="border: none; border-top: 1.2px solid lightgray; width: 100%;">

### 2. Docking protocol 
 For the control we will use the exact protocol as for the apo docking just with the new input files we just edited before.

      # ====================================================================
      #   Docking of ens 4 AF +1 bound protein and ens 4 simrna +1 bound RNA
      # ====================================================================
      
      # directory name of the run
      run_dir = "control_apo_vs_bound"
      
      # local mode
      mode = "local"
      ncores = 50
      # Self contained rundir (to avoid problems with long filename paths)
      self_contained = true
      
      # molecules to be docked
      molecules =  [
      "af_4_p_1.pdb",
      "simrna_4_r_1.pdb"
      ]
      
      # ====================================================================
      # Parameters for the various stages
      # ====================================================================
      [topoaa]

      [rigidbody]
      # surface ambig restraints
      ambig_fname = "pesto_residues_80.tbl"
      randremoval = False
      # Number of models to generate
      sampling = 10000
      
      [clustfcc]
      min_population = 1
      
      [seletopclusts]
      ## select all the clusters
      top_cluster = 10
      ## select the best 10 models of each cluster
      top_models = 10
      
      [caprieval]
      allatoms = True
      
      [flexref]
      # Acceptable percentage of model failures
      tolerance = 5
      ambig_fname = "pesto_residues_80.tbl"
      randremoval = False
      dnarest_on=True
      
      [emref]
      # surface ambig restraints
      ambig_fname = "pesto_residues_80.tbl"
      randremoval = False
      
      [clustfcc]
      
      [seletopclusts]
      top_cluster = 10
      
      [caprieval]
      allatoms = True
      
      [caprieval]
      reference_fname = "reference_5.pdb"
      allatoms = True
      
      # ====================================================================

<hr style="border: none; border-top: 1.2px solid lightgray; width: 100%;">

### 3.Analysis
When evaluating the docking results, our goal is to determine whether the bound conformations of the protein–RNA complex are represented among the top-ranked models, and whether rigid-body docking successfully captured shape complementarity between the binding partners.

We begin by examining the CAPRI table to identify the highest-performing models, then use the traceback file to trace each model back to its corresponding input structure. The traceback file can be found here:

    control_apo_vs_bound/traceback/traceback.tsv
<br>
<img src="/figures/course/capri_control.png" width="75%">
<img src="/figures/course/control_plot.png" width="75%">
<img src="/figures/course/control_plot_2.png" width="75%">
<br>

As we can see, the protein aligns very well with the bound structure, but the RNA does not adopt the correct bound conformation. This suggests that the docking approach may lack the flexibility needed to account for the conformational changes required for binding—especially on the RNA side. The image shows the first-ranked docked model superimposed with the first reference model.

<br>
<img src="/figures/course/control_apo_vs_bound.png" width="75%">
<br>

<hr style="border: none; border-top: 2px solid lightgray; width: 100%;">

### Conclusions
* **Protein–RNA Complex Challenges:** Protein–RNA interactions often involve diverse and cooperative binding modes—particularly in systems with multiple RNA Recognition Motifs (RRMs)—that may deviate from classical recognition patterns.

* **Impact of Protein Conformations:** The input protein structure strongly influences docking results. Certain conformations are better suited to the RNA binding interface, while others lead to weaker or less accurate interactions.

* **RNA Flexibility:** Single-stranded RNA can undergo significant conformational changes upon binding. Capturing this behavior is challenging for rigid-body docking protocols, especially when refinement time is limited.

* **Advantages of Modular Workflows:** Combining tools like PESTO with tailored HADDOCK3 modules allows for more efficient, targeted docking setups and can lead to better prediction outcomes.
